torch/_higher_order_ops/wrap.py

```diff
@@ -9,5 +10,23 @@ def __call__(self, func, *args):
         result = func(*args)
         return result
 
-
 wrap = Wrap()
+
+class WrapActivationCheckpoint(HigherOrderOperator):
+    """
+    Wrap activation checkpoint is little different from Wrap. There are two
+    functions to wrap here - utils.checkpoint() itself and the function to be
+    checkpointed (the first arg to the utils.checkpoint() function).
+    """
+    def __init__(self):
+        super().__init__("wrap_activation_checkpoint")
+
+    def __call__(self, function, *args, **kwargs):
+        # use_reentrant is set to False because this op is going to be traced.
+        # And we ensure that AOT Autograd traces through the non reentrant
+        # version of checkpointing.
+        kwargs["use_reentrant"] = False
+        result = checkpoint(function, *args, **kwargs)
```

Comment:
Can you add a test where this line throws? I am wondering if we handle this correctly.

Reply:
+1, curious about this

Reply:
Hmm, not sure what we are checking here?

~~~
        def gn(x, y):
            raise RuntimeError("Injecting error")
            return torch.sigmoid(torch.matmul(x, y))

        def fn(x, y):
            return torch.cos(torch.utils.checkpoint.checkpoint(gn, torch.sin(x), y))

        x = torch.randn(4, 4, requires_grad=True)
        y = torch.randn(4, 4, requires_grad=True)
        args = (x, y)

        backend = EagerAndRecordGraphs()
        cnt = CompileCounterWithBackend(backend)

        # expected = fn(*args)
        result = torch.compile(fn, backend=cnt)(*args)
~~~

Are we checking something like this raises the same error for each and torch.compile?

Reply:
Let's say the function being checkpointed raises an exception. There are two cases:
1. Dynamo graph breaks on exceptions. Then we would expect checkpoint(func, *args) to go the fallback route.
2. Dynamo preserves exceptions in the graph (seems unlikely to me, there's no try-catch syntax in FX). Then we would expect checkpoint(func, *args) to not fallback to eager mode, but still raise an exception later down the line (maybe in AOTAutograd?)

So yes, we would expect an error to be raised for eager and torch.compile. Though I'm more curious about which of the two above cases we're in.